sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp

```diff
@@ -121,91 +143,532 @@ struct SptrsvWrap {
+
+    // multiply_subtract. C -= A * B
+    KOKKOS_INLINE_FUNCTION
+    static void multiply_subtract(const scalar_t &A, const scalar_t &B,
+                                  scalar_t &C) {
+      C -= A * B;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    static void copy(const member_type &, scalar_t &, const scalar_t &) {}
+
+    KOKKOS_INLINE_FUNCTION
+    static void copy(scalar_t &, const scalar_t &) {}
+
+    // lget
+    KOKKOS_INLINE_FUNCTION
+    scalar_t &lget(const size_type row) const { return lhs(row); }
+
+    // rget
+    KOKKOS_INLINE_FUNCTION
+    scalar_t rget(const size_type row) const { return rhs(row); }
+
+    // vget
+    KOKKOS_INLINE_FUNCTION
+    scalar_t vget(const size_type nnz) const { return values(nnz); }
+
+    // lhs = (lhs + rhs) / diag (team)
+    KOKKOS_INLINE_FUNCTION
+    static void add_and_divide(const member_type &team, scalar_t &lhs_val,
+                               const scalar_t &rhs_val,
+                               const scalar_t &diag_val) {
+      Kokkos::single(Kokkos::PerTeam(team),
+                     [&]() { lhs_val = (lhs_val + rhs_val) / diag_val; });
+    }
+
+    // lhs = (lhs + rhs) / diag (serial)
+    KOKKOS_INLINE_FUNCTION
+    static void add_and_divide(scalar_t &lhs_val, const scalar_t &rhs_val,
+                               const scalar_t &diag_val) {
+      lhs_val = (lhs_val + rhs_val) / diag_val;
+    }
+
+    // print
+    KOKKOS_INLINE_FUNCTION
+    static void print(const scalar_t &item) { std::cout << item << std::endl; }
+
+    KOKKOS_INLINE_FUNCTION
+    static void print(ArrayType rhs, const int) {
+      std::cout << rhs << std::endl;
+    }
+  };
+
+  // Partial specialization for block support
+  template <class RowMapType, class EntriesType, class ValuesType,
+            class LHSType, class RHSType>
+  struct Common<RowMapType, EntriesType, ValuesType, LHSType, RHSType, true> {
+    // BSR data is in LayoutRight!
+    using Layout = Kokkos::LayoutRight;
+
+    using Block = Kokkos::View<
+        scalar_t **, Layout, typename ValuesType::device_type,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    // const block
+    using CBlock = Kokkos::View<
+        const scalar_t **, Layout, typename ValuesType::device_type,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    // scratch block
+    using SBlock = Kokkos::View<
+        scalar_t **, Layout, typename execution_space::scratch_memory_space,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    using Vector = Kokkos::View<
+        scalar_t *, Layout, typename ValuesType::device_type,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    using CVector = Kokkos::View<
+        const scalar_t *, Layout, typename ValuesType::device_type,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    static constexpr size_type MAX_VEC_SIZE = 11;
```

Comment:
Why is this 11, is there an algorithmic reason or an architectural one?
It seems strange that the same number would work well for both CPU and GPU?

Reply:
I understand the confusion. The issue is that I wanted BUFF_SIZE=128 for the thread-local temporary buffer used in `divide`. Since the buffer is for a temporary block, that means the biggest dimension that won't overflow the buffer is 11 (11*11 < 128). I could bump BUFF_SIZE to 256 easily.

Reply:
Does that mean that this code will just throw if the block size is > 11? It should work for all block sizes, even if it's optimized for small block sizes

Reply:
Yes, it will throw if block_size is > MAX_VEC_SIZE. I definitely do not want to a heap allocation here and block_size is not known at compile time, so I think some static limit is needed. Maybe there's an approach I didn't think of.

Reply:
I guess the other approach would be PerThread scratch?

Reply:
Actually what Luc said below would be reasonable. Keep the statically allocated limit, but loop over chunks of K rows at a time, where K is the statically allocated block size limit (could be 11, or something else).

Another way might be to not do a single vector-valued reduction, and instead have one thread-level reduction per row. That would probably improve performance too, but it would also require more code changes from what you have now